src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -24,38 +24,54 @@
+//                                    the false-path loop.
 //
-// orig:                       transformed:
-//                               if (invariant-test) then
-//  predicates                     predicates
-//  loop                           loop
-//    stmt1                          stmt1
-//    if (invariant-test) then       stmt2
-//      stmt2                        stmt4
-//    else                         endloop
-//      stmt3                    else
-//    endif                        predicates [clone]
-//    stmt4                        loop [clone]
-//  endloop                          stmt1 [clone]
-//                                   stmt3
-//                                   stmt4 [clone]
-//                                 endloop
-//                               endif
+// The invariant, non-loop-exiting test can now be moved before both loops (to only execute it once) and turned into a
+// loop selector If node to select at runtime which unswitched loop version should be executed.
+// - Loop selector true?  Execute the true-path-loop.
+// - Loop selector false? Execute the false-path-loop.
 //
-// Note: the "else" clause may be empty
-
-//------------------------------policy_unswitching-----------------------------
-// Return TRUE or FALSE if the loop should be unswitched
-// (ie. clone loop with an invariant test that does not exit the loop)
-bool IdealLoopTree::policy_unswitching( PhaseIdealLoop *phase ) const {
+// Note that even though an invariant test that exits the loop could also be optimized with Loop Unswitching, it is more
+// efficient to simply peel the loop which achieves the same result in a simpler manner (also see policy_peeling()).
+//
+// The following graph summarizes the Loop Unswitching optimization:
+//
+//                                                                    [Initialized Assertion Predicates]
+//                                                                                    |
+//    [Predicates]                                                         loop selector If (invariant-test)
+//         |                                                                  /                   \
+//    Original Loop                                                       true?                  false?
+//      stmt1                                                             /                         \
+//      if (invariant-test)           UNSWITCHED              [Cloned Parse Predicates]         [Cloned Parse Predicates]
+//        if-block                    =========>              [Cloned Template                  [Cloned Template
+//      else                                                   Assertion Predicates]             Assertion Predicates]
+//        // could be empty                                         |                                  |
+//        [else-block]                                        True-Path-Loop                    False-Path-Loop
```

Comment:
Could it also be the other way around, that the if-path is empty but the else-path has code?

Reply:
Yes definitely, but this is more a reflection of Java code where you always have an if block but not always an else block. But it might be cleaner to just remove the "could be empty" comment and make it explicit and not worry about whether one block is empty. I'll do that. The image should just give a hint :-)